File: app/src/main/java/com/lessonscontrol/activities/ViewStudentActivity.java
package com.lessonscontrol.activities;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.widget.ImageButton;
import android.widget.TextView;

import com.lessonscontrol.adapter.LessonListAdapter;
import com.lessonscontrol.data.entities.Lesson;
import com.lessonscontrol.data.entities.Student;
import com.lessonscontrol.data.viewModel.LessonViewModel;

import java.security.InvalidParameterException;
import java.util.List;


public class ViewStudentActivity extends AppCompatActivity {

    /**
     * Request code for this activity. This code will be returned when the activity exits.
     * @see android.support.v4.app.FragmentActivity#startActivityForResult
     */
    public static final int VIEW_STUDENT_ACTIVITY_REQUEST_CODE = 2;

    private Student student;

    private LessonViewModel lessonViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_view_student);
        this.lessonViewModel = ViewModelProviders.of(this).get(LessonViewModel.class);

        this.student = this.getIntent().getParcelableExtra(Student.STUDENT_EXTRA_KEY);

        if (this.student == null) {
            String invalidParameterMessage = "Student should not be null on ViewStudentActivity.";
            Log.e(this.getClass().toString(), invalidParameterMessage);
            throw new InvalidParameterException(invalidParameterMessage);
        }

        this.populateActivityWithStudentInfo();

        RecyclerView lessonsRecyclerView = findViewById(R.id.lessons_recycler_view);
        final LessonListAdapter lessonListAdapter = new LessonListAdapter(this);
        lessonsRecyclerView.setAdapter(lessonListAdapter);
        lessonsRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        this.observeDataChange(lessonListAdapter);

        ImageButton editStudentButton = findViewById(R.id.button_edit_student);
        editStudentButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent editStudentActivityIntent = new Intent(ViewStudentActivity.this, EditStudentActivity.class);
                editStudentActivityIntent.putExtra(Student.STUDENT_EXTRA_KEY, student);
                ViewStudentActivity.this.startActivityForResult(editStudentActivityIntent, EditStudentActivity.EDIT_STUDENT_ACTIVITY_REQUEST_CODE);
            }
        });

        ImageButton addLessonButton = findViewById(R.id.button_add_lesson);
        addLessonButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent newLessonIntent = new Intent(ViewStudentActivity.this, EditLessonActivity.class);
                newLessonIntent.putExtra(EditLessonActivity.STUDENT_ID_KEY, student.getID());
                ViewStudentActivity.this.startActivityForResult(newLessonIntent, EditLessonActivity.EDIT_LESSON_ACTIVITY_REQUEST_CODE);
            }
        });
    }

    private void observeDataChange(final LessonListAdapter lessonListAdapter) {
        this.lessonViewModel.findLessonsByStudent(this.student).observe(this, new Observer<List<Lesson>>() {
            @Override
            public void onChanged(@Nullable List<Lesson> lessons) {
                lessonListAdapter.setLessons(lessons);
            }
        });
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        //Updates the screen if the EditStudentActivity returns success (student updated)
        if (requestCode == EditStudentActivity.EDIT_STUDENT_ACTIVITY_REQUEST_CODE && resultCode == RESULT_OK) {
            this.student = data.getParcelableExtra(Student.STUDENT_EXTRA_KEY);
            this.populateActivityWithStudentInfo();
        }
    }

    private void populateActivityWithStudentInfo() {
        ((TextView) findViewById(R.id.view_name)).setText(student.getName());
        ((TextView) findViewById(R.id.view_phone)).setText(student.getPhone());
        ((TextView) findViewById(R.id.view_mail)).setText(student.getEmail());
        ((TextView) findViewById(R.id.view_address)).setText(student.getAddress());
    }
}

File: app/src/main/java/com/lessonscontrol/activities/ViewStudentActivity.kt
package com.lessonscontrol.activities

import android.app.Activity
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.widget.ImageButton
import com.lessonscontrol.activities.dataHandler.ActivityDataHandlerFactory
import com.lessonscontrol.activities.dataHandler.impl.ViewStudentActivityDataHandler
import com.lessonscontrol.adapter.LessonListAdapter
import com.lessonscontrol.data.entities.Student
import com.lessonscontrol.data.viewModel.LessonViewModel
import com.lessonscontrol.data.viewModel.StudentViewModel

class ViewStudentActivity : AppCompatActivity() {

    lateinit var lessonListAdapter: LessonListAdapter
        private set

    lateinit var student: Student
        private set

    lateinit var lessonViewModel: LessonViewModel
        private set

    lateinit var studentViewModel: StudentViewModel
        private set

    private val dataHandler = ActivityDataHandlerFactory.createDataHandler(this) as ViewStudentActivityDataHandler

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_view_student)

        lessonViewModel = ViewModelProviders.of(this).get(LessonViewModel::class.java)
        studentViewModel = ViewModelProviders.of(this).get(StudentViewModel::class.java)
        student = intent.getParcelableExtra(Student.STUDENT_EXTRA_KEY)

        dataHandler.populateFields()

        val lessonsRecyclerView = findViewById<RecyclerView>(R.id.lessons_recycler_view)
        lessonListAdapter = LessonListAdapter(this)
        lessonsRecyclerView.adapter = lessonListAdapter
        lessonsRecyclerView.layoutManager = LinearLayoutManager(this)
        dataHandler.observeDataChange()

        val editStudentButton = findViewById<ImageButton>(R.id.button_edit_student)
        editStudentButton.setOnClickListener {
            val editStudentActivityIntent = Intent(this@ViewStudentActivity, EditStudentActivity::class.java)
            editStudentActivityIntent.putExtra(Student.STUDENT_EXTRA_KEY, student)
            this@ViewStudentActivity.startActivityForResult(editStudentActivityIntent, EditStudentActivity.EDIT_STUDENT_ACTIVITY_REQUEST_CODE)
        }

        val deleteStudentButton = findViewById<ImageButton>(R.id.button_delete_student)
        deleteStudentButton.setOnClickListener {
            AlertDialog.Builder(this@ViewStudentActivity)
                    .setTitle(resources.getString(R.string.delete_student))
                    .setPositiveButton(android.R.string.yes) { _, _ ->
                        studentViewModel.delete(student)
                        finish()
                    }.setNegativeButton(android.R.string.no, null).show()
        }

        val addLessonButton = findViewById<ImageButton>(R.id.button_add_lesson)
        addLessonButton.setOnClickListener {
            val newLessonIntent = Intent(this@ViewStudentActivity, EditLessonActivity::class.java)
            newLessonIntent.putExtra(EditLessonActivity.STUDENT_ID_KEY, student.id)
            this@ViewStudentActivity.startActivityForResult(newLessonIntent, EditLessonActivity.EDIT_LESSON_ACTIVITY_REQUEST_CODE)
        }
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)

        //Updates the screen if the EditStudentActivity returns success (student updated)
        if (requestCode == EditStudentActivity.EDIT_STUDENT_ACTIVITY_REQUEST_CODE && resultCode == Activity.RESULT_OK) {
            student = data.getParcelableExtra(Student.STUDENT_EXTRA_KEY)
            dataHandler.populateFields()
        }
    }

    companion object {
        /**
         * Request code for this activity. This code will be returned when the activity exits.
         *
         * @see android.support.v4.app.FragmentActivity.startActivityForResult
         */
        const val VIEW_STUDENT_ACTIVITY_REQUEST_CODE = 2
    }
}
